File: Platforms/Common/Reign_Audio_XAudio_Component/Audio.cpp
#include "pch.h"
#include "Audio.h"

namespace Reign_Audio_XAudio_Component
{
	#pragma region Constructors
	AudioErrors AudioCom::Init()
	{
		null();
		
		IXAudio2* xAudioTEMP;
		if(FAILED(XAudio2Create(&xAudioTEMP, 0)))
		{
			return AudioErrors::XAudio2;
		}
		xAudio = xAudioTEMP;

		IXAudio2MasteringVoice* masteringVoiceTEMP;
		if (FAILED(xAudio->CreateMasteringVoice(&masteringVoiceTEMP)))
		{
			return AudioErrors::MasteringVoice;
		}
		masteringVoice = masteringVoiceTEMP;

		return AudioErrors::None;
	}

	AudioCom::~AudioCom()
	{
		if (masteringVoice) masteringVoice->DestroyVoice();
		if (xAudio) xAudio->Release();
		null();
	}

	void AudioCom::null()
	{
		masteringVoice = 0;
		xAudio = 0;
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Audio.XAudio/Audio.cs
using Reign.Core;
using Reign_Audio_XAudio_Component;
using System;

namespace Reign.Audio.XAudio
{
	public class Audio : Disposable, AudioI
	{
		#region Properties
		internal delegate void UpdateCallbackFunc();
		internal UpdateCallbackFunc UpdateCallback;

		internal AudioCom com;
		#endregion

		#region Constructors
		public Audio(DisposableI parent)
		: base(parent)
		{
			try
			{
				com = new AudioCom();
				var error = com.Init();

				switch (error)
				{
					case AudioErrors.XAudio2: Debug.ThrowError("Audio", "Failed to create XAudio2"); break;
					case AudioErrors.MasteringVoice: Debug.ThrowError("Audio", "Failed to create MasteringVoice"); break;
				}
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Update()
		{
			if (UpdateCallback != null) UpdateCallback();
		}
		#endregion
	}
}
